### PR TITLE
DKG: ban misbehaving operators from sortition-pool rewards

### PR DESCRIPTION
**Pull request by tomaszslabon**
This PR modifies `random-beacon` so that the group members that misbehaved during DKG are punished (banned from the `sortition-pools` rewards for some time):
* misbehaved members are stored for the candidate group when a DKG result is submitted
* when the DKG result is approved, the stored members are punished(banned from the sortition-pools rewards for some period of time)

There are two small TODOs that depend on other PRs, but I think we could already merge this PR.

### CODE REVIEW

**Comment by pdyraga** [comment]
Note that as discussed in #2671 review, we can't remove operators inactive during the relay entry signing from the pool because the pool is locked. Instead, we are going to ban them from sortition pool rewards. We could theoretically remove inactive DKG operators from the pool because the pool is getting unlocked when DKG result is approved but I do not think it is worth it. Instead, I'd use the same solution mainly because of gas efficiency problems we have to solve, and remove them from rewards instead.

File: solidity/random-beacon/contracts/libraries/Groups.sol
Comment (lukasz-zimnoch) [inline solidity/random-beacon/contracts/libraries/Groups.sol:175]: If we move it to `DKG.sol` we can have just `dkg.submittedResultMisbehavedMembers`.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Done in https://github.com/keep-network/keep-core/pull/2700/commits/e1ddb2aeb68d23ec7b2c44904eebd2febc979f0d.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/RandomBeacon.sol:508]: Yes, I think it can be reused.
If #2707 is merged first, I will update this parameter.
If this PR is merged first, then we can merge it later (I suggest we keep the TODO for now).

File: solidity/random-beacon/contracts/libraries/Groups.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/Groups.sol]: It makes sense. Done in https://github.com/keep-network/keep-core/pull/2700/commits/cf0a8164ec36c74b8dec05201c10d6c4e6ec4d74.

**Comment by pdyraga** [comment]
@tomaszslabon Can you please update PR title and description?

**Comment by tomaszslabon** [comment]
> @tomaszslabon Can you please update PR title and description?

Sure. Done.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:507]: I just approved #2707, should get merged once all checks pass.

Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:507]: Let's use the same value - `sortitionPoolRewardsBanDuration`.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/RandomBeacon.sol:555]: @pdyraga How do we go about this one?
I don't think we should handle this TODO in this PR, but make a task and handle it in another PR.

File: solidity/random-beacon/contracts/libraries/DKG.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: I did something like this:
```
        uint32[] memory misbehavedMembers = new uint32[](result.misbehavedMembersIndices.length);
        for (uint256 i = 0; i < result.misbehavedMembersIndices.length; i++) {
            // group member indices start from 1, so we need to -1 on misbehaved
            uint32 memberArrayPosition = result.misbehavedMembersIndices[i] - 1;
            misbehavedMembers[i] = result.members[memberArrayPosition];
        }
        self.submittedResultMisbehavedMembers = misbehavedMembers;
```

Unfortunately, it increase the average consumption: 1186454 (before) vs 1189217 (after) for `submitDkgResult()` :(

Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: And when I put an `if` around it:
```
        if (result.misbehavedMembersIndices.length > 0) {
            uint32[] memory misbehavedMembers = new uint32[](result.misbehavedMembersIndices.length);
            for (uint256 i = 0; i < result.misbehavedMembersIndices.length; i++) {
                // group member indices start from 1, so we need to -1 on misbehaved
                uint32 memberArrayPosition = result.misbehavedMembersIndices[i] - 1;
                misbehavedMembers[i] = result.members[memberArrayPosition];
            }
            self.submittedResultMisbehavedMembers = misbehavedMembers;
        }
 ```
 
 it fell to 1186517, which is still slightly higher than the original.

Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: In what conditions you were testing it? No misbehaving members / 5 misbehaving members / 64 misbehaving? Average gas consumption from all unit tests might not be the most accurate number. I think in the vast majority of cases we should have from to ~5 misbehaving members and that's what we should optimize for.

But still, it's possible that the current version will be better for this case, just wanted to confirm.

Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: I just ran all the tests from `RandomBeacon.GroupCreation.test.ts` and there is only one test with misbehaved members (3 indices).
I will comment out the other tests and increase the number to 5.

Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: Now I commented out every test expect [this one](https://github.com/keep-network/keep-core/blob/022d12020922493d1a90ef5d6731c6fa49ed113f/solidity/random-beacon/test/RandomBeacon.GroupCreation.test.ts#L1248) and changed the  misbehaved to  [2, 10, 18, 37, 64].

The consumption is very high: 1312275 (without local memory array) vs 1313839 (with local memory array).
BTW, I realised we should have another test for `submitDkgResult` were we test if the misbehaved members are set.

Comment (tomaszslabon) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: Now I tried the same with no misbehaved:
1243317 (without local array) vs 1243311 (with local array), so there is almost no difference.

Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/DKG.sol:225]: Interestiiing... OK, let's leave it as-is.

**Comment by pdyraga** [comment]
From @tomaszslabon:
> BTW, I realized we should have another test for submitDkgResult where we test if the misbehaved members are set.

Can we add this test in this PR? I am good to 🚢 once it is done.

File: solidity/random-beacon/contracts/RandomBeacon.sol
Comment (tomaszslabon) [reply solidity/random-beacon/contracts/RandomBeacon.sol:555]: The cost of `approveDkgResult` is:
* `142930` if there are no misbehaved members
* `186578` when there are 5

It will go higher when we implement the sortition-pool's banFromRewards function

**Comment by tomaszslabon** [comment]
> From @tomaszslabon:
> 
> > BTW, I realized we should have another test for submitDkgResult where we test if the misbehaved members are set.
> 
> Can we add this test in this PR? I am good to 🚢 once it is done.

@pdyraga 
Done in https://github.com/keep-network/keep-core/pull/2700/commits/1847dfa901064e2f2ddc14e1c8e5b091ae9fd7ba.
We must also add a task, like:
"DKG: Pay submitter's reward to anyone if the submitter does not approve the result on time"

**Review by pdyraga — APPROVED** [review]

**Pull request merged**
